### DISCUSSION TITLE
Support for JWTs using RS256

### DISCUSSION THREAD

**Discussion by davidmuzi** [supabase/supabase#4059] [Questions]
When using the OAuth based flow for Sign in with Apple, `https://abc.supabase.co/auth/v1/authorize?provider=apple&redirect_to=myapp://auth"` returns a JWT signed using HS256

JWT header
```
{
  "alg": "HS256",
  "typ": "JWT"
}
```

When using a [SignInWithAppleButton](https://developer.apple.com/documentation/authenticationservices/signinwithapplebutton) in an iOS app, a RS256 JWT is returned.

```swift
struct LoginView: View {
    var body: some View {
        VStack {
            SignInWithAppleButton(.signIn) { request in
                request.requestedScopes = [.fullName, .email]
            } onCompletion: { result in
                switch result {
                    case .success(let authResults):
                    guard let credentials = authResults.credential as? ASAuthorizationAppleIDCredential, let identityToken = credentials.identityToken, let identityTokenString = String(data: identityToken, encoding: .utf8) else { return }
                    print(identityTokenString)
                    case .failure(let error):
                        print("Authorization failed: \(error.localizedDescription)")
                }
            }
        }
    }
}
```

which has the header
```
{
  "kid": "YuyXoY",
  "alg": "RS256"
}
```

 when if used results in
```
{
  "code": 401,
  "msg": "Invalid token: signing method RS256 is invalid"
}
```

Is it possible to use support this token, or must the OAuth flow be used to get an HS256 JWT?

**Comment by mstade** [comment]
The JWT tokens that can be used are signed with HS256 using the JWT secret in the Supabase dashboard. Even if you somehow get Apple to send you a token signed with HS256 which is a so called symmetric algorithm (you need the secret to verify it) as opposed to RS256 which is an asymmetric algorithm (signed with a private key, can be verified with a public key) you'd still need it to be signed with the Supabase JWT secret. I don't know if you can actually use the `SignInWithAppleButton` API to authenticate with Supabase, you'd have to kick off the Supabase flow instead I think, which probably isn't ideal in an iOS app since it means having to use a web view, I think.

#3899 has a few nuggets of information that may be useful to you.

**Reply by mstade** [reply]
FWIW I'd very much be in favor of support for asymmetric signing algorithms in Supabase! 🙏🏻

**Comment by J0** [comment]
We're looking into Asymmetric JWTs :)

**Reply by Manouchehri** [reply]
Having JWE built-in would be quite nice too.

**Comment by wiverson** [comment]
FWIW I worked with the nice folks at gotrue-csharp to get the native sign in with Apple flow working.

https://github.com/supabase-community/gotrue-csharp

There were two different flows, one for PKCE and one for Apple that were required.

https://github.com/supabase-community/gotrue-csharp/blob/c1bf5c14f87f3c0305fbe954dc6c410f65b4f810/Gotrue/Helpers.cs#L48
vs
https://github.com/supabase-community/gotrue-csharp/blob/c1bf5c14f87f3c0305fbe954dc6c410f65b4f810/Gotrue/Helpers.cs#L64

These were cribbed off the dart GoTrue implementation for flutter.

I made a video covering this from a high level, https://www.youtube.com/watch?v=S0hTwtsUWcM with links to some useful gists in the description.
